### PR TITLE
Use MOCs for alignment and generate MOC in from_dataframe

### PR DESCRIPTION
**Pull request by smcguire-cmu**
Uses the MOC alignment functions recently added to hipscat to perform alignment for crossmatch and join with higher order coverage. Generates a MOC when `from_dataframe` is called.

### CODE REVIEW

File: src/lsdb/dask/merge_catalog_functions.py [hunk 1]
```diff
@@ -54,25 +56,51 @@ def concat_partition_and_margin(
...
 
-def align_catalogs(left: Catalog, right: Catalog) -> PixelAlignment:
+def align_catalogs(left: Catalog, right: Catalog, add_right_margin: bool = True) -> PixelAlignment:
     """Aligns two catalogs, also using the right catalog's margin if it exists
 
     Args:
         left (lsdb.Catalog): The left catalog to align
         right (lsdb.Catalog): The right catalog to align
+        add_right_margin (float): When using MOCs to align catalogs, adds a border to the
+            right catalog's moc to include the margin of the right catalog, if the right margin exists
```
Comment (camposandro) [inline src/lsdb/dask/merge_catalog_functions.py:66]:
```suggestion
        add_right_margin (bool): If True, when using MOCs to align catalogs, adds a border to the
            right catalog's moc to include the margin of the right catalog, if it exists. Defaults to True.
```

File: src/lsdb/dask/merge_catalog_functions.py [hunk 2]
```diff
@@ -54,25 +56,51 @@ def concat_partition_and_margin(
...
+        if right.hc_structure.moc is not None
+        else right.hc_structure.pixel_tree.to_moc()
+    )
+    if right_added_radius is not None:
+        right_moc_depth_resol = hp.nside2resol(hp.order2nside(right_moc.max_order), arcmin=True) * 60
+        if right_added_radius < right_moc_depth_resol:
+            right_moc = right_moc.add_neighbours()
+        else:
+            delta_order = int(np.ceil(np.log2(right_added_radius / right_moc_depth_resol)))
+            right_moc = right_moc.degrade_to_order(right_moc.max_order - delta_order).add_neighbours()
```
Comment (camposandro) [inline src/lsdb/dask/merge_catalog_functions.py:95]: Could we move this auxiliary code to hipscat?

**Review by camposandro — COMMENTED** [review]
This looks awesome!

**Comment by camposandro** [comment]
Could you also run a join using MOCs for one of those cases where not specifying a high lower order on `from_dataframe` would just render the workflow unfeasible? It'd be great to see how this change speeds up the computation. If I recall correctly [this](https://github.com/lincc-frameworks/notebooks_lf/blob/main/DELVE_gaia_xmatch/Spectroscopy.ipynb) was one of the workflows running into issues!

File: src/lsdb/dask/merge_catalog_functions.py (hunk 2 above)
Comment (smcguire-cmu) [reply src/lsdb/dask/merge_catalog_functions.py:95]: Problem is hipscat catalog doesn't have the margin as a parameter. We should refactor a lot more stuff into hipscat I think, so let's open up a ticket to do that.

Comment (smcguire-cmu) [reply src/lsdb/dask/merge_catalog_functions.py:95]: opened #325

**Review by camposandro — APPROVED** [review]

**Pull request merged**